src/software/world/ball.cpp

```diff
@@ -1,32 +1,41 @@
 #include "software/world/ball.h"
 
 #include "shared/constants.h"
+#include "software/world/ball_state.h"
 
 Ball::Ball(Point position, Vector velocity, const Timestamp &timestamp,
            unsigned int history_duration)
-    : positions_(history_duration),
-      velocities_(history_duration),
-      last_update_timestamps(history_duration)
+    : states_(history_duration)
 {
-    addStateToBallHistory(position, velocity, timestamp);
+    updateCurrentState(position, velocity, timestamp);
 }
 
-void Ball::updateState(const Ball &new_ball_data)
+Ball::Ball(BallState &ball_state, unsigned int history_duration)
+    : states_(history_duration)
 {
-    updateState(new_ball_data.position(), new_ball_data.velocity(),
-                new_ball_data.lastUpdateTimestamp());
+    states_.push_front(ball_state);
 }
 
-void Ball::updateState(const Point &new_position, const Vector &new_velocity,
-                       const Timestamp &timestamp)
+BallState Ball::currentState() const
 {
-    if (timestamp < lastUpdateTimestamp())
+    return states_.front();
+}
+
+void Ball::updateCurrentState(const BallState &new_state)
+{
+    if (new_state.timestamp() < lastUpdateTimestamp() && !states_.empty())
```

Comment:
Is it necessary to check if `states_` is empty? Seems like `lastUpdateTimestamp` calls `front()`, which asserts that buffer isn't empty. If the function should take care of the case where the buffer is empty and so it shouldn't do a comparison, I think the empty check should come first.

Reply:
It's necessary because the comparison new_state.timestamp() < lastUpdateTimestamp() returns true when the states_ buffer is empty for some reason

Even though it shouldn't technically matter, good call about lastUpdateTimestamp() being called before the empty check. I'll change the order.